In [27]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [28]:
import sqlite3
import pandas as pd
db_path = "/content/drive/MyDrive/BIO_MEDICAL_PROJECT/obesity_gene_analysis/data/OBESITYGENE_PROJECT.db"

In [105]:
x_columns = [

    # metadata 테이블
    'treatment', 'sex', 'age_yrs', 'height_cm'
    'base_weight_kg', 'base_bmi_kg_m2', 'base_body_fat_pct', 'base_pure_fat_kg',

    # weight_bmi_changes 테이블
    'weight_kg_rate_AB', 'weight_kg_rate_BC', 'weight_kg_rate_AC',
    'bmi_rate_AB', 'bmi_rate_BC', 'bmi_rate_AC',
    'fat_rate_AB', 'fat_rate_BC', 'fat_rate_AC',
    'pure_fat_rate_AB', 'pure_fat_rate_BC', 'pure_fat_rate_AC',

    # expression_change 테이블
    'exp_rate_AB', 'exp_rate_BC', 'exp_rate_AC',
]

In [31]:
conn = sqlite3.connect(db_path)

platform_df=pd.read_sql("SELECT * FROM platform", conn)
metadata_df=pd.read_sql("SELECT * FROM metadata", conn)
weight_bmi_changes_df=pd.read_sql("SELECT * FROM weight_bmi_changes", conn)
expression_change_df=pd.read_sql("SELECT * FROM expression_change", conn)

conn.close()

In [32]:
platform_df.head()

ID                                            GB_LIST  \
0  7896736                                               None   
1  7896738                                               None   
2  7896740  NM_001005240,NM_001004195,NM_001005484,BC13684...   
3  7896742                                  BC118988,AL137655   
4  7896744  NM_001005277,NM_001005221,NM_001005224,NM_0010...   

              SPOT_ID seqname      RANGE_GB RANGE_STRAND  RANGE_START  \
0    chr1:53049-54936    chr1  NC_000001.10            +        53049   
1    chr1:63015-63887    chr1  NC_000001.10            +        63015   
2    chr1:69091-70008    chr1  NC_000001.10            +        69091   
3  chr1:334129-334296    chr1  NC_000001.10            +       334129   
4  chr1:367659-368597    chr1  NC_000001.10            +       367659   

   RANGE_STOP  total_probes  \
0       54936             7   
1       63887            31   
2       70008            24   
3      334296             6   
4      368597            36   

                                     gene_assignment  \
0                                                ---   
1                                                ---   
2  NM_001005240 // OR4F17 // olfactory receptor, ...   
3  ENST00000388975 // SEPT14 // septin 14 // 7p11...   
4  NM_001005277 // OR4F16 // olfactory receptor, ...   

                                     mrna_assignment category gene_symbol  
0                                                ---      ---        None  
1  ENST00000328113 // ENSEMBL // cdna:pseudogene ...     main        None  
2  NM_001005240 // RefSeq // Homo sapiens olfacto...     main      OR4F17  
3  ENST00000388975 // ENSEMBL // Septin-14 gene:E...     main      SEPT14  
4  NM_001005277 // RefSeq // Homo sapiens olfacto...     main      OR4F16

In [33]:
metadata_df

gsm_id subject_id                                tissue  \
0    GSM2062466          1  abdominal subcutaneous white adipose   
1    GSM2062467          1  abdominal subcutaneous white adipose   
2    GSM2062468          1  abdominal subcutaneous white adipose   
3    GSM2062469          2  abdominal subcutaneous white adipose   
4    GSM2062470          2  abdominal subcutaneous white adipose   
..          ...        ...                                   ...   
147  GSM2062613         56  abdominal subcutaneous white adipose   
148  GSM2062614         59  abdominal subcutaneous white adipose   
149  GSM2062615         59  abdominal subcutaneous white adipose   
150  GSM2062616         61  abdominal subcutaneous white adipose   
151  GSM2062617         61  abdominal subcutaneous white adipose   

                 treatment                  time_point     sex  age_yrs  \
0    very-low-calorie diet              at study start  female     46.0   
1    very-low-calorie diet    after weight loss period  female     46.0   
2    very-low-calorie diet  after weight stable period  female     46.0   
3    very-low-calorie diet              at study start  female     45.0   
4    very-low-calorie diet    after weight loss period  female     45.0   
..                     ...                         ...     ...      ...   
147       low-calorie diet  after weight stable period    male     56.0   
148       low-calorie diet              at study start    male     67.0   
149       low-calorie diet    after weight loss period    male     67.0   
150  very-low-calorie diet              at study start    male     52.0   
151  very-low-calorie diet  after weight stable period    male     52.0   

     height_cm  weight_kg  bmi_kg_m2  body_fat_pct  pure_fat_kg  \
0        167.0      83.38      29.90          39.9     33.26862   
1        167.0      75.16      26.95          35.8     26.90728   
2        167.0      73.79      26.46          32.4     23.90796   
3        164.0      80.94      30.09          46.7     37.79898   
4        164.0      71.92      26.74          43.0     30.92560   
..         ...        ...        ...           ...          ...   
147      171.0      89.98      30.77          36.5     32.84270   
148      168.5      82.42      29.03          28.2     23.24244   
149      168.5      75.01      26.42          21.2     15.90212   
150      182.0      94.26      28.46          31.1     29.31486   
151      182.0      88.50      26.72          26.6     23.54100   

     base_weight_kg  base_bmi_kg_m2  base_body_fat_pct  base_pure_fat_kg  
0             83.38           29.90               39.9          33.26862  
1             83.38           29.90               39.9          33.26862  
2             83.38           29.90               39.9          33.26862  
3             80.94           30.09               46.7          37.79898  
4             80.94           30.09               46.7          37.79898  
..              ...             ...                ...               ...  
147           93.36           31.93               40.0          37.34400  
148           82.42           29.03               28.2          23.24244  
149           82.42           29.03               28.2          23.24244  
150           94.26           28.46               31.1          29.31486  
151           94.26           28.46               31.1          29.31486  

[152 rows x 16 columns]

왜 weight_bmi_changes테이블의 rate열이 채워지지 않았는지 추후 재실험을 통해 원인을 찾고 싶음. 아쉽게 결과지표의 rate열은 모델에 쓰이지 못했음.

In [34]:
weight_bmi_changes_df.head()

subject_id  weight_kg_C_A  weight_kg_C_B  weight_kg_B_A  bmi_C_A  bmi_C_B  \
0          1          -9.59          -1.37          -8.22    -3.44    -0.49   
1          2          -9.97          -0.95          -9.02    -3.70    -0.35   
2          3          -9.47           2.13         -11.60    -2.97     0.68   
3          4          -4.91           0.11          -5.02    -2.07     0.05   
4          5         -12.65          -0.35         -12.30    -3.95    -0.11   

   bmi_B_A  fat_C_A  fat_C_B  fat_B_A  ...  weight_kg_rate_AC  bmi_rate_AB  \
0    -2.95     -7.5     -3.4     -4.1  ...               None         None   
1    -3.35     -6.2     -2.5     -3.7  ...               None         None   
2    -3.65     -6.4      0.3     -6.7  ...               None         None   
3    -2.12     -2.5     -0.2     -2.3  ...               None         None   
4    -3.84     -8.9     -1.9     -7.0  ...               None         None   

   bmi_rate_BC bmi_rate_AC  fat_rate_AB fat_rate_BC fat_rate_AC  \
0      -0.1225        None         None      -0.850        None   
1      -0.0875        None         None      -0.625        None   
2       0.1700        None         None       0.075        None   
3       0.0125        None         None      -0.050        None   
4      -0.0275        None         None      -0.475        None   

   pure_fat_rate_AB pure_fat_rate_BC pure_fat_rate_AC  
0              None        -0.749830             None  
1              None        -0.545687             None  
2              None         0.176775             None  
3              None        -0.025560             None  
4              None        -0.421758             None  

[5 rows x 25 columns]

In [35]:
expression_change_df.head()

ID_REF subject_id   exp_C_A   exp_C_B   exp_B_A exp_rate_AB  exp_rate_BC  \
0  7892501          1 -0.402629 -0.149945 -0.252684        None    -0.037486   
1  7892502          1  1.251101  0.883600  0.367501        None     0.220900   
2  7892503          1 -1.115590 -0.330360 -0.785230        None    -0.082590   
3  7892504          1  0.849925  0.743890  0.106035        None     0.185973   
4  7892505          1 -1.044279  0.261170 -1.305449        None     0.065293   

  exp_rate_AC  
0        None  
1        None  
2        None  
3        None  
4        None

In [36]:
vlcd_subjects = metadata_df[metadata_df['treatment'] == 'very-low-calorie diet']['subject_id'].unique().tolist()
lcd_subjects = metadata_df[metadata_df['treatment'] == 'low-calorie diet']['subject_id'].unique().tolist()

In [37]:
weight_bmi_changes_df.columns

Index(['subject_id', 'weight_kg_C_A', 'weight_kg_C_B', 'weight_kg_B_A',
       'bmi_C_A', 'bmi_C_B', 'bmi_B_A', 'fat_C_A', 'fat_C_B', 'fat_B_A',
       'pure_fat_C_A', 'pure_fat_C_B', 'pure_fat_B_A', 'weight_kg_rate_AB',
       'weight_kg_rate_BC', 'weight_kg_rate_AC', 'bmi_rate_AB', 'bmi_rate_BC',
       'bmi_rate_AC', 'fat_rate_AB', 'fat_rate_BC', 'fat_rate_AC',
       'pure_fat_rate_AB', 'pure_fat_rate_BC', 'pure_fat_rate_AC'],
      dtype='object')

In [38]:
print(len(vlcd_subjects))
print(len(lcd_subjects))

26
27


In [39]:
# VLCD, LCD 주 수 설정
vlcd_weeks = 5
lcd_weeks = 12

# 함수 정의: 주 수 반환
def get_weeks(subject_id):
    if subject_id in vlcd_subjects:
        return vlcd_weeks
    elif subject_id in lcd_subjects:
        return lcd_weeks
    else:
        return np.nan  # 알 수 없는 treatment일 경우

# 각 subject별 주 수 컬럼 추가
weight_bmi_changes_df['weeks_AB'] = weight_bmi_changes_df['subject_id'].apply(get_weeks)

# 새로운 비율 컬럼 계산
weight_bmi_changes_df['weight_kg_rate_AB'] = weight_bmi_changes_df['weight_kg_B_A'] / weight_bmi_changes_df['weeks_AB']
weight_bmi_changes_df['bmi_rate_AB'] = weight_bmi_changes_df['bmi_B_A'] / weight_bmi_changes_df['weeks_AB']
weight_bmi_changes_df['fat_rate_AB'] = weight_bmi_changes_df['fat_B_A'] / weight_bmi_changes_df['weeks_AB']
weight_bmi_changes_df['pure_fat_rate_AB'] = weight_bmi_changes_df['pure_fat_B_A'] / weight_bmi_changes_df['weeks_AB']

# 주 수: AB 구간 + 4주 = AC 구간
weeks_AC = weight_bmi_changes_df['weeks_AB'] + 4

# rate_AC 계산
weight_bmi_changes_df['bmi_rate_AC'] = weight_bmi_changes_df['bmi_C_A'] / weeks_AC
weight_bmi_changes_df['fat_rate_AC'] = weight_bmi_changes_df['fat_C_A'] / weeks_AC
weight_bmi_changes_df['pure_fat_rate_AC'] = weight_bmi_changes_df['pure_fat_C_A'] / weeks_AC

In [40]:
weight_bmi_changes_df

subject_id  weight_kg_C_A  weight_kg_C_B  weight_kg_B_A  bmi_C_A  bmi_C_B  \
0           1          -9.59          -1.37          -8.22    -3.44    -0.49   
1           2          -9.97          -0.95          -9.02    -3.70    -0.35   
2           3          -9.47           2.13         -11.60    -2.97     0.68   
3           4          -4.91           0.11          -5.02    -2.07     0.05   
4           5         -12.65          -0.35         -12.30    -3.95    -0.11   
5           6          -3.94           1.87          -5.81    -1.57     0.75   
6           8          -5.66           0.58          -6.24    -2.08     0.21   
7           9          -9.94          -0.57          -9.37    -3.46    -0.20   
8          10         -10.48          -1.02          -9.46    -3.55    -0.35   
9          11          -9.73          -0.73          -9.00    -3.00    -0.22   
10         12          -5.14           0.97          -6.11    -1.89     0.36   
11         13         -11.14          -0.02         -11.12    -3.91    -0.01   
12         14         -10.08           0.67         -10.75    -2.93     0.20   
13         15            NaN          -0.19            NaN      NaN    -0.07   
14         18         -10.42            NaN            NaN    -3.15      NaN   
15         19          -3.93           0.36          -4.29    -1.63     0.15   
16         20          -9.44          -0.96          -8.48    -3.29    -0.34   
17         21          -4.67           0.24          -4.91    -1.86     0.09   
18         22          -5.65           4.06          -9.71    -1.63     1.17   
19         23         -11.85          -0.13         -11.72    -3.74    -0.04   
20         24         -10.14          -0.63          -9.51    -3.57    -0.22   
21         25          -8.41          -0.53          -7.88    -2.73    -0.17   
22         26         -12.16          -1.30         -10.86    -3.42    -0.37   
23         27          -7.62          -0.55          -7.07    -2.65    -0.19   
24         28          -8.92          -2.13          -6.79    -3.34    -0.80   
25         29         -11.34          -2.12          -9.22    -3.28    -0.61   
26         30          -9.31           0.40          -9.71    -2.68     0.11   
27         31          -5.82          -0.88          -4.94    -2.16    -0.33   
28         33          -6.70          -0.46          -6.24    -2.07    -0.14   
29         34          -9.51          -0.80          -8.71    -3.76    -0.32   
30         35          -6.38           0.35          -6.73    -2.22     0.12   
31         36         -10.35          -1.18          -9.17    -3.58    -0.41   
32         37          -8.18          -0.40          -7.78    -2.89    -0.14   
33         38         -11.86          -1.52         -10.34    -4.12    -0.53   
34         39          -3.68           0.40          -4.08    -1.35     0.15   
35         40          -8.43          -1.23          -7.20    -3.38    -0.49   
36         41          -8.67           0.33          -9.00    -3.28     0.13   
37         42         -12.87          -0.18         -12.69    -3.95    -0.05   
38         43         -11.72          -0.16         -11.56    -3.52    -0.05   
39         45          -7.61          -0.35          -7.26    -2.79    -0.13   
40         46         -11.69          -2.27          -9.42    -3.88    -0.75   
41         47          -8.64          -1.46          -7.18    -2.98    -0.50   
42         48          -8.66           0.43          -9.09    -2.64     0.13   
43         49         -10.64          -0.96          -9.68    -3.39    -0.30   
44         50          -4.33           0.06          -4.39    -1.42     0.02   
45         51          -8.82          -0.64          -8.18    -2.85    -0.21   
46         52          -8.13           1.81          -9.94    -2.54     0.57   
47         53            NaN            NaN          -9.28      NaN      NaN   
48         54            NaN          -0.41            NaN      NaN    -0.11   
49 

In [41]:
# 주차 정보 가져오기
weeks_df = weight_bmi_changes_df[['subject_id', 'weeks_AB']].copy()
weeks_df['weeks_AC'] = weeks_df['weeks_AB'] + 4

# subject_id 기준으로 병합 (join)
expression_change_df = expression_change_df.merge(weeks_df, on='subject_id', how='left')

# 계산
expression_change_df['exp_rate_AB'] = expression_change_df['exp_B_A'] / expression_change_df['weeks_AB']
expression_change_df['exp_rate_AC'] = expression_change_df['exp_C_A'] / expression_change_df['weeks_AC']

In [42]:
expression_change_df

ID_REF subject_id   exp_C_A   exp_C_B   exp_B_A  exp_rate_AB  \
0        7892501          1 -0.402629 -0.149945 -0.252684    -0.050537   
1        7892502          1  1.251101  0.883600  0.367501     0.073500   
2        7892503          1 -1.115590 -0.330360 -0.785230    -0.157046   
3        7892504          1  0.849925  0.743890  0.106035     0.021207   
4        7892505          1 -1.044279  0.261170 -1.305449    -0.261090   
...          ...        ...       ...       ...       ...          ...   
1764736  8180414         61 -0.522933       NaN       NaN          NaN   
1764737  8180415         61 -0.023247       NaN       NaN          NaN   
1764738  8180416         61  0.040307       NaN       NaN          NaN   
1764739  8180417         61 -0.013860       NaN       NaN          NaN   
1764740  8180418         61 -0.794944       NaN       NaN          NaN   

         exp_rate_BC  exp_rate_AC  weeks_AB  weeks_AC  
0          -0.037486    -0.044737         5         9  
1           0.220900     0.139011         5         9  
2          -0.082590    -0.123954         5         9  
3           0.185973     0.094436         5         9  
4           0.065293    -0.116031         5         9  
...              ...          ...       ...       ...  
1764736          NaN    -0.058104         5         9  
1764737          NaN    -0.002583         5         9  
1764738          NaN     0.004479         5         9  
1764739          NaN    -0.001540         5         9  
1764740          NaN    -0.088327         5         9  

[1764741 rows x 10 columns]

In [75]:
meta_base = metadata_df[metadata_df['time_point'] == 'at study start'].copy()
meta_base = meta_base[['subject_id', 'treatment', 'sex', 'age_yrs',
                       'base_weight_kg', 'base_bmi_kg_m2', 'base_body_fat_pct', 'base_pure_fat_kg']]

In [76]:
import pandas as pd
model1_shap_df=pd.read_csv("/content/drive/MyDrive/BIO_MEDICAL_PROJECT/obesity_gene_analysis/data/shap1_df.csv", index_col=0)

index_col=0

'Unnamed: 0' 열을 인덱스로 설정하고 불러오기


In [77]:
model1_shap_df.columns

Index(['HNRNPA1P10', 'BTN2A1', 'PTPRD', 'SNORD116-1', 'CDCA5', 'C8orf80',
       'ARF5', 'ANP32B', 'SULT1C4', 'CBY1',
       ...
       'ABAT', 'TRO', 'SFMBT2', 'SRCIN1', 'ADAMTS4', 'ABHD12B', 'ZNF705D',
       'AFAP1L2', 'DLX3', 'ATP1A3'],
      dtype='object', length=138)

In [78]:
model1_shap_df

HNRNPA1P10    BTN2A1     PTPRD  SNORD116-1     CDCA5   C8orf80  \
weight_kg_C_A    0.247319  0.103196  0.077105    0.201080  0.092313  0.009061   
bmi_C_A          0.062802  0.018025  0.022044    0.044329  0.017000  0.004891   
fat_C_A          0.315680  0.183632  0.079298    0.160215  0.053380  0.013559   
pure_fat_C_A     0.431585  0.172992  0.069219    0.334195  0.146440  0.048397   

                   ARF5    ANP32B   SULT1C4      CBY1  ...      ABAT  \
weight_kg_C_A  0.053822  0.012425  0.027417  0.038274  ...  0.004355   
bmi_C_A        0.011383  0.001588  0.008996  0.009588  ...  0.000105   
fat_C_A        0.059069  0.287340  0.037754  0.103328  ...  0.001175   
pure_fat_C_A   0.086344  0.167967  0.091685  0.078083  ...  0.011762   

                    TRO    SFMBT2    SRCIN1   ADAMTS4   ABHD12B   ZNF705D  \
weight_kg_C_A  0.002476  0.040468  0.014481  0.000937  0.007409  0.222945   
bmi_C_A        0.000466  0.027148  0.001102  0.001640  0.000412  0.035223   
fat_C_A        0.000232  0.013709  0.002584  0.003000  0.001980  0.002301   
pure_fat_C_A   0.000474  0.048596  0.000120  0.001375  0.000016  0.117619   

                AFAP1L2      DLX3    ATP1A3  
weight_kg_C_A  0.000758  0.006048  0.002923  
bmi_C_A        0.000250  0.000079  0.006358  
fat_C_A        0.002685  0.029798  0.002862  
pure_fat_C_A   0.002408  0.002202  0.007281  

[4 rows x 138 columns]

In [79]:
# SHAP 유전자 리스트를 대문자로 변환
gene_list = model1_shap_df.columns.str.upper().unique().tolist()

# platform_df의 유전자명 대문자로 변환
platform_df['gene_symbol_upper'] = platform_df['gene_symbol'].str.upper()

# expression_change_df['ID_REF']와 platform_df['ID']를 기준으로 병합
expr_merged = expression_change_df.merge(platform_df[['ID', 'gene_symbol_upper']],
                                         left_on='ID_REF', right_on='ID')

# SHAP 유전자만 필터링
filtered_expr_change = expr_merged[expr_merged['gene_symbol_upper'].isin(gene_list)].copy()

In [80]:
filtered_expr_change

ID_REF subject_id   exp_C_A   exp_C_B   exp_B_A  exp_rate_AB  \
4307     7897955          1  0.132273  0.698371 -0.566099    -0.113220   
4415     7898858          1 -0.236056  0.134038 -0.370094    -0.074019   
4520     7899870          1 -0.277461 -0.091510 -0.185950    -0.037190   
5195     7906475          1 -0.049847 -0.132065  0.082218     0.016444   
6118     7916282          1  0.400305  0.012280  0.388024     0.077605   
...          ...        ...       ...       ...       ...          ...   
1763272  8169291         61 -0.005153       NaN       NaN          NaN   
1763361  8170135         61  0.146637       NaN       NaN          NaN   
1763729  8173430         61 -0.486585       NaN       NaN          NaN   
1764016  8175690         61  0.333193       NaN       NaN          NaN   
1764353  8178554         61 -0.345111       NaN       NaN          NaN   

         exp_rate_BC  exp_rate_AC  weeks_AB  weeks_AC       ID  \
4307        0.174593     0.014697         5         9  7897955   
4415        0.033509    -0.026228         5         9  7898858   
4520       -0.022878    -0.030829         5         9  7899870   
5195       -0.033016    -0.005539         5         9  7906475   
6118        0.003070     0.044478         5         9  7916282   
...              ...          ...       ...       ...      ...   
1763272          NaN    -0.000573         5         9  8169291   
1763361          NaN     0.016293         5         9  8170135   
1763729          NaN    -0.054065         5         9  8173430   
1764016          NaN     0.037021         5         9  8175690   
1764353          NaN    -0.038346         5         9  8178554   

        gene_symbol_upper  
4307              AADACL4  
4415             C1ORF213  
4520               ZNF362  
5195                FCRL6  
6118                 LRP8  
...                   ...  
1763272             ATG4A  
1763361            GPR112  
1763729             SNX12  
1764016           MAGEA10  
1764353            LY6G5C  

[7632 rows x 12 columns]

In [81]:
# 유전자 발현값 wide 형식으로 변환
gene_df = filtered_expr_change[['ID_REF', 'subject_id', 'exp_C_A', 'exp_C_B', 'exp_B_A',
       'exp_rate_AB', 'exp_rate_BC', 'exp_rate_AC']]
pivot_gene_df = gene_df.pivot(index='subject_id', columns='ID_REF', values=['exp_C_A', 'exp_C_B', 'exp_B_A',
       'exp_rate_AB', 'exp_rate_BC', 'exp_rate_AC'])

In [82]:
gene_df

ID_REF subject_id   exp_C_A   exp_C_B   exp_B_A  exp_rate_AB  \
4307     7897955          1  0.132273  0.698371 -0.566099    -0.113220   
4415     7898858          1 -0.236056  0.134038 -0.370094    -0.074019   
4520     7899870          1 -0.277461 -0.091510 -0.185950    -0.037190   
5195     7906475          1 -0.049847 -0.132065  0.082218     0.016444   
6118     7916282          1  0.400305  0.012280  0.388024     0.077605   
...          ...        ...       ...       ...       ...          ...   
1763272  8169291         61 -0.005153       NaN       NaN          NaN   
1763361  8170135         61  0.146637       NaN       NaN          NaN   
1763729  8173430         61 -0.486585       NaN       NaN          NaN   
1764016  8175690         61  0.333193       NaN       NaN          NaN   
1764353  8178554         61 -0.345111       NaN       NaN          NaN   

         exp_rate_BC  exp_rate_AC  
4307        0.174593     0.014697  
4415        0.033509    -0.026228  
4520       -0.022878    -0.030829  
5195       -0.033016    -0.005539  
6118        0.003070     0.044478  
...              ...          ...  
1763272          NaN    -0.000573  
1763361          NaN     0.016293  
1763729          NaN    -0.054065  
1764016          NaN     0.037021  
1764353          NaN    -0.038346  

[7632 rows x 8 columns]

In [84]:
pivot_gene_df.shape

(53, 864)

In [85]:
pivot_gene_df

exp_C_A                                                    \
ID_REF       7897955   7898858   7899870   7906475   7916282   7919314   
subject_id                                                               
1           0.132273 -0.236056 -0.277461 -0.049847  0.400305  0.000487   
10         -0.362438  0.288727 -0.040643 -0.144525  0.115295 -0.041263   
11         -0.329518  0.042324 -0.031332 -0.333983  0.085071  0.213110   
12         -0.609585 -0.088545  0.134744  0.330162  0.070191  0.086244   
13         -0.226196  0.091607 -0.055057 -0.584748 -0.539106  0.143447   
14         -0.271020 -0.286017 -0.293426  0.333576  0.479702  0.124518   
15               NaN       NaN       NaN       NaN       NaN       NaN   
18         -0.175578 -0.138242 -0.270724 -0.188614 -0.257367 -0.293513   
19         -0.403363  0.431593  0.348716 -0.580449  0.126299  0.039207   
2           0.052973 -0.592131 -0.179559  0.498617  0.276134 -0.074901   
20          0.551088  0.077000  0.277833  0.004603  0.110233  0.092335   
21          0.186254  0.053312  0.553460  0.018997  0.185984  0.420057   
22         -0.156781 -0.861017 -0.057083  0.428779 -0.076878  0.493167   
23         -0.015601  0.195180  0.109662  0.219995  0.185909 -0.559521   
24          0.039508  0.031731  0.076123  0.188697 -0.368058  0.169756   
25          0.203050  0.059686  0.090034 -0.307978 -0.144536  0.338490   
26         -0.014917 -0.255120 -0.141872  0.017112  0.112708 -0.310464   
27         -0.012916 -0.406930  0.030669 -0.453836  0.413515 -0.284127   
28          0.372658  0.382743 -0.048437 -0.203270 -0.154321 -0.628808   
29          0.429388  0.108003  0.089199  0.740068 -0.155644 -0.425963   
3          -0.202976  0.314981 -0.053662  0.184861 -0.115912  0.002334   
30          0.015143 -0.348865  0.153818 -0.670036  0.095927 -0.370572   
31         -0.399931 -0.168273  0.211743  0.667028  0.361972  0.479684   
33         -0.085691  0.043209 -0.294254 -0.397314 -0.340595  0.066535   
34         -0.059939 -0.049996 -0.334454  0.456729  0.101104  0.582783   
35         -0.130508  0.000759 -0.119665 -0.478262  0.170311  0.068770   
36         -0.190438 -0.187449 -0.105579  0.052766 -0.094543 -0.153977   
37          0.434692 -0.234401 -0.141412 -0.414913  0.150478 -0.116612   
38         -0.278343 -0.408507  0.098474 -0.379407 -0.005578  0.178524   
39          0.688899 -0.039550 -0.003385  0.438824  0.215388 -0.667099   
4          -0.368213  0.253984 -0.006555  0.396088 -0.348209 -0.016100   
40         -0.322482 -0.327682 -0.365889  0.256739  0.089087  0.085621   
41          0.051462 -0.098192 -0.100844  0.098126  0.025049 -0.005971   
42         -0.440767  0.455804  0.395719 -0.557789 -0.238760 -0.357153   
43         -0.052786  0.035065  0.349600 -0.521480 -0.285767 -0.296735   
45          0.358288  0.421074 -0.228348  0.382414 -0.293948  0.273462   
46          0.056407 -0.142148 -0.113537 -0.420464 -0.512316  0.274289   
47          0.161448 -0.298472 -0.237604 -0.254267  0.053569  0.446417   
48          0.303239 -0.171928 -0.176747 -0.525579 -0.572819 -0.229201   
49          0.124528  0.424867  0.310741  0.331061 -0.235071 -0.010002   
5          -0.657459  0.482898  0.423221  1.110574 -0.187989  0.195969   
50         -0.239018  0.295736  0.171553  0.441593  0.024273  0.285402   
51          0.278464 -0.205783  0.361291 -0.431840  0.043680 -0.738405   
52          0.170166  0.358066 -0.044276 -0.170924  0.188394 -0.162508   
53               NaN       NaN       NaN       NaN       NaN       NaN   
54               NaN       NaN       NaN       NaN       NaN       NaN   
55          0.076785  0.080032  0.369600 -0.341776 -0.248913 -0.071061   
56         -0.025414  0.469843 -0.185468  0.053802  0.074283 -0.335575   
59               NaN       NaN       NaN       NaN       NaN       NaN   
6          -0.340253 -0.138042 -0.049597  0.599317 -0.326102 -0.140729   
61          0.177950  0.104126  0.003109 -0.059764  0.341721  0.066961   
8           0.398

len(gene_list)는 138인데, pivot_gene_df의 행의 수를 피실험자 수인 53으로 나누면 144이 된다-> 즉, 유전자가 144개라는 뜻.

6개 더 많아졌는데 아마 multi-index와 관련이 있다고 생각하지만, 정확하게 원인을 밝혀내지 못해 추후 재실험을 통해 규명하고 싶음!

In [98]:
X=pivot_gene_df

In [97]:
pivot_gene_df.shape

(53, 864)

In [93]:
cols_to_exclude = [
    'weight_kg_rate_AB', 'weight_kg_rate_BC', 'weight_kg_rate_AC',
    'bmi_rate_AB', 'bmi_rate_BC', 'bmi_rate_AC',
    'fat_rate_AB', 'fat_rate_BC', 'fat_rate_AC',
    'pure_fat_rate_AB', 'pure_fat_rate_BC', 'pure_fat_rate_AC',
    'weeks_AB'
]

weight_bmi_changes_filtered = weight_bmi_changes_df.drop(columns=cols_to_exclude)

In [95]:
weight_bmi_changes_filtered.shape

(53, 13)

In [94]:
weight_bmi_changes_filtered

subject_id  weight_kg_C_A  weight_kg_C_B  weight_kg_B_A  bmi_C_A  bmi_C_B  \
0           1          -9.59          -1.37          -8.22    -3.44    -0.49   
1           2          -9.97          -0.95          -9.02    -3.70    -0.35   
2           3          -9.47           2.13         -11.60    -2.97     0.68   
3           4          -4.91           0.11          -5.02    -2.07     0.05   
4           5         -12.65          -0.35         -12.30    -3.95    -0.11   
5           6          -3.94           1.87          -5.81    -1.57     0.75   
6           8          -5.66           0.58          -6.24    -2.08     0.21   
7           9          -9.94          -0.57          -9.37    -3.46    -0.20   
8          10         -10.48          -1.02          -9.46    -3.55    -0.35   
9          11          -9.73          -0.73          -9.00    -3.00    -0.22   
10         12          -5.14           0.97          -6.11    -1.89     0.36   
11         13         -11.14          -0.02         -11.12    -3.91    -0.01   
12         14         -10.08           0.67         -10.75    -2.93     0.20   
13         15            NaN          -0.19            NaN      NaN    -0.07   
14         18         -10.42            NaN            NaN    -3.15      NaN   
15         19          -3.93           0.36          -4.29    -1.63     0.15   
16         20          -9.44          -0.96          -8.48    -3.29    -0.34   
17         21          -4.67           0.24          -4.91    -1.86     0.09   
18         22          -5.65           4.06          -9.71    -1.63     1.17   
19         23         -11.85          -0.13         -11.72    -3.74    -0.04   
20         24         -10.14          -0.63          -9.51    -3.57    -0.22   
21         25          -8.41          -0.53          -7.88    -2.73    -0.17   
22         26         -12.16          -1.30         -10.86    -3.42    -0.37   
23         27          -7.62          -0.55          -7.07    -2.65    -0.19   
24         28          -8.92          -2.13          -6.79    -3.34    -0.80   
25         29         -11.34          -2.12          -9.22    -3.28    -0.61   
26         30          -9.31           0.40          -9.71    -2.68     0.11   
27         31          -5.82          -0.88          -4.94    -2.16    -0.33   
28         33          -6.70          -0.46          -6.24    -2.07    -0.14   
29         34          -9.51          -0.80          -8.71    -3.76    -0.32   
30         35          -6.38           0.35          -6.73    -2.22     0.12   
31         36         -10.35          -1.18          -9.17    -3.58    -0.41   
32         37          -8.18          -0.40          -7.78    -2.89    -0.14   
33         38         -11.86          -1.52         -10.34    -4.12    -0.53   
34         39          -3.68           0.40          -4.08    -1.35     0.15   
35         40          -8.43          -1.23          -7.20    -3.38    -0.49   
36         41          -8.67           0.33          -9.00    -3.28     0.13   
37         42         -12.87          -0.18         -12.69    -3.95    -0.05   
38         43         -11.72          -0.16         -11.56    -3.52    -0.05   
39         45          -7.61          -0.35          -7.26    -2.79    -0.13   
40         46         -11.69          -2.27          -9.42    -3.88    -0.75   
41         47          -8.64          -1.46          -7.18    -2.98    -0.50   
42         48          -8.66           0.43          -9.09    -2.64     0.13   
43         49         -10.64          -0.96          -9.68    -3.39    -0.30   
44         50          -4.33           0.06          -4.39    -1.42     0.02   
45         51          -8.82          -0.64          -8.18    -2.85    -0.21   
46         52          -8.13           1.81          -9.94    -2.54     0.57   
47         53            NaN            NaN          -9.28      NaN      NaN   
48         54            NaN          -0.41            NaN      NaN    -0.11   
49 

In [90]:
y = weight_bmi_changes_df.set_index('subject_id')

In [92]:
weight_bmi_changes_df.columns

Index(['subject_id', 'weight_kg_C_A', 'weight_kg_C_B', 'weight_kg_B_A',
       'bmi_C_A', 'bmi_C_B', 'bmi_B_A', 'fat_C_A', 'fat_C_B', 'fat_B_A',
       'pure_fat_C_A', 'pure_fat_C_B', 'pure_fat_B_A', 'weight_kg_rate_AB',
       'weight_kg_rate_BC', 'weight_kg_rate_AC', 'bmi_rate_AB', 'bmi_rate_BC',
       'bmi_rate_AC', 'fat_rate_AB', 'fat_rate_BC', 'fat_rate_AC',
       'pure_fat_rate_AB', 'pure_fat_rate_BC', 'pure_fat_rate_AC', 'weeks_AB'],
      dtype='object')

In [99]:
print(X.shape)
print(y.shape)

(53, 864)
(53, 25)


In [119]:
pivot_gene_df.shape

(53, 864)

In [111]:
print(metadata_df['time_point'].unique())

['at study start' 'after weight loss period' 'after weight stable period']


In [114]:
# 필요한 컬럼만 선택
weight_cols = [
    'subject_id', 'weight_kg_rate_AB', 'weight_kg_rate_BC', 'weight_kg_rate_AC',
    'bmi_rate_AB', 'bmi_rate_BC', 'bmi_rate_AC',
    'fat_rate_AB', 'fat_rate_BC', 'fat_rate_AC',
    'pure_fat_rate_AB', 'pure_fat_rate_BC', 'pure_fat_rate_AC'
]
meta_cols = [
    'subject_id', 'treatment', 'sex', 'age_yrs', 'height_cm',
    'base_weight_kg', 'base_bmi_kg_m2', 'base_body_fat_pct', 'base_pure_fat_kg'
]

# 각각 필요한 컬럼만 추출
weight_df = weight_bmi_changes_df[weight_cols].copy()
meta_df = metadata_df[metadata_df['time_point'] == 'at study start'][meta_cols].copy()

# subject_id 기준으로 순차적으로 병합
merged_df = weight_df.merge(meta_df, on='subject_id')

# 4. subject_id를 인덱스로 설정
final_merged_df=merged_df.set_index('subject_id')

In [118]:
meta_df.shape

(51, 9)

In [117]:
weight_df.shape

(53, 13)

In [116]:
final_merged_df.shape

(51, 20)

In [ ]:
pivot_gene_df

In [120]:
# 공통 인덱스 (subject_id) 추출
common_index = final_merged_df.index.intersection(pivot_gene_df.index)

# 공통 인덱스 기준으로 X, y 정렬 및 필터링
final_merged_df_common = final_merged_df.loc[common_index]
pivot_gene_df_common = pivot_gene_df.loc[common_index]

In [121]:
print(final_merged_df_common.shape)
print(pivot_gene_df_common.shape)
print(final_merged_df_common.index.equals(pivot_gene_df_common.index))  # True여야 안전

(51, 20)
(51, 864)
True


In [126]:
pivot_gene_df_common

exp_C_A                                                    \
ID_REF       7897955   7898858   7899870   7906475   7916282   7919314   
subject_id                                                               
1           0.132273 -0.236056 -0.277461 -0.049847  0.400305  0.000487   
2           0.052973 -0.592131 -0.179559  0.498617  0.276134 -0.074901   
3          -0.202976  0.314981 -0.053662  0.184861 -0.115912  0.002334   
4          -0.368213  0.253984 -0.006555  0.396088 -0.348209 -0.016100   
5          -0.657459  0.482898  0.423221  1.110574 -0.187989  0.195969   
6          -0.340253 -0.138042 -0.049597  0.599317 -0.326102 -0.140729   
8           0.398210  0.252762 -0.280807 -0.016487 -0.204304  0.290873   
9           0.614803  0.070689  0.014549  0.216768  0.113467  0.220625   
10         -0.362438  0.288727 -0.040643 -0.144525  0.115295 -0.041263   
11         -0.329518  0.042324 -0.031332 -0.333983  0.085071  0.213110   
12         -0.609585 -0.088545  0.134744  0.330162  0.070191  0.086244   
13         -0.226196  0.091607 -0.055057 -0.584748 -0.539106  0.143447   
14         -0.271020 -0.286017 -0.293426  0.333576  0.479702  0.124518   
18         -0.175578 -0.138242 -0.270724 -0.188614 -0.257367 -0.293513   
19         -0.403363  0.431593  0.348716 -0.580449  0.126299  0.039207   
20          0.551088  0.077000  0.277833  0.004603  0.110233  0.092335   
21          0.186254  0.053312  0.553460  0.018997  0.185984  0.420057   
22         -0.156781 -0.861017 -0.057083  0.428779 -0.076878  0.493167   
23         -0.015601  0.195180  0.109662  0.219995  0.185909 -0.559521   
24          0.039508  0.031731  0.076123  0.188697 -0.368058  0.169756   
25          0.203050  0.059686  0.090034 -0.307978 -0.144536  0.338490   
26         -0.014917 -0.255120 -0.141872  0.017112  0.112708 -0.310464   
27         -0.012916 -0.406930  0.030669 -0.453836  0.413515 -0.284127   
28          0.372658  0.382743 -0.048437 -0.203270 -0.154321 -0.628808   
29          0.429388  0.108003  0.089199  0.740068 -0.155644 -0.425963   
30          0.015143 -0.348865  0.153818 -0.670036  0.095927 -0.370572   
31         -0.399931 -0.168273  0.211743  0.667028  0.361972  0.479684   
33         -0.085691  0.043209 -0.294254 -0.397314 -0.340595  0.066535   
34         -0.059939 -0.049996 -0.334454  0.456729  0.101104  0.582783   
35         -0.130508  0.000759 -0.119665 -0.478262  0.170311  0.068770   
36         -0.190438 -0.187449 -0.105579  0.052766 -0.094543 -0.153977   
37          0.434692 -0.234401 -0.141412 -0.414913  0.150478 -0.116612   
38         -0.278343 -0.408507  0.098474 -0.379407 -0.005578  0.178524   
39          0.688899 -0.039550 -0.003385  0.438824  0.215388 -0.667099   
40         -0.322482 -0.327682 -0.365889  0.256739  0.089087  0.085621   
41          0.051462 -0.098192 -0.100844  0.098126  0.025049 -0.005971   
42         -0.440767  0.455804  0.395719 -0.557789 -0.238760 -0.357153   
43         -0.052786  0.035065  0.349600 -0.521480 -0.285767 -0.296735   
45          0.358288  0.421074 -0.228348  0.382414 -0.293948  0.273462   
46          0.056407 -0.142148 -0.113537 -0.420464 -0.512316  0.274289   
47          0.161448 -0.298472 -0.237604 -0.254267  0.053569  0.446417   
48          0.303239 -0.171928 -0.176747 -0.525579 -0.572819 -0.229201   
49          0.124528  0.424867  0.310741  0.331061 -0.235071 -0.010002   
50         -0.239018  0.295736  0.171553  0.441593  0.024273  0.285402   
51          0.278464 -0.205783  0.361291 -0.431840  0.043680 -0.738405   
52          0.170166  0.358066 -0.044276 -0.170924  0.188394 -0.162508   
53               NaN       NaN       NaN       NaN       NaN       NaN   
55          0.076785  0.080032  0.369600 -0.341776 -0.248913 -0.071061   
56         -0.025414  0.469843 -0.185468  0.053802  0.074283 -0.335575   
59               NaN       NaN       NaN       NaN       NaN       NaN   
61          0.177950  0.104126  0.003109 -0.059764  0.341721  0.066961   

                

In [127]:
# 멀티 컬럼 인덱스 → 단일 컬럼 인덱스 문자열로 변환
pivot_gene_df_common.columns = ['{}_{}'.format(a, b) for a, b in pivot_gene_df_common.columns]

In [128]:
# 인덱스가 동일한지 다시 확인 (필요시 정렬)
pivot_gene_df_common = pivot_gene_df_common.sort_index()
final_merged_df_common = final_merged_df_common.sort_index()

In [131]:
pivot_gene_df_common.shape

(51, 864)

In [132]:
final_merged_df_common.shape

(51, 20)

In [133]:
# 3. 병합 (subject_id를 인덱스로 맞춘 후 inner join)
X = final_merged_df_common.join(pivot_gene_df_common, how='inner')

In [134]:
X

weight_kg_rate_AB  weight_kg_rate_BC weight_kg_rate_AC  \
subject_id                                                           
1                   -1.644000            -0.3425              None   
10                  -0.788333            -0.2550              None   
11                  -0.750000            -0.1825              None   
12                  -1.222000             0.2425              None   
13                  -0.926667            -0.0050              None   
14                  -0.895833             0.1675              None   
18                        NaN                NaN              None   
19                  -0.357500             0.0900              None   
2                   -1.804000            -0.2375              None   
20                  -0.706667            -0.2400              None   
21                  -0.982000             0.0600              None   
22                  -1.942000             1.0150              None   
23                  -0.976667            -0.0325              None   
24                  -0.792500            -0.1575              None   
25                  -1.576000            -0.1325              None   
26                  -2.172000            -0.3250              None   
27                  -1.414000            -0.1375              None   
28                  -1.358000            -0.5325              None   
29                  -1.844000            -0.5300              None   
3                   -2.320000             0.5325              None   
30                  -1.942000             0.1000              None   
31                  -0.411667            -0.2200              None   
33                  -0.520000            -0.1150              None   
34                  -0.725833            -0.2000              None   
35                  -0.560833             0.0875              None   
36                  -0.764167            -0.2950              None   
37                  -1.556000            -0.1000              None   
38                  -2.068000            -0.3800              None   
39                  -0.340000             0.1000              None   
4                   -1.004000             0.0275              None   
40                  -1.440000            -0.3075              None   
41                  -0.750000             0.0825              None   
42                  -2.538000            -0.0450              None   
43                  -2.312000            -0.0400              None   
45                  -1.452000            -0.0875              None   
46                  -1.884000            -0.5675              None   
47                  -1.436000            -0.3650              None   
48                  -0.757500             0.1075              None   
49                  -0.806667            -0.2400              None   
5                   -2.460000            -0.0875              None   
50                  -0.365833             0.0150              None   
51                  -0.681667            -0.1600              None   
52                  -1.988000             0.4525              None   
53                  -0.773333                NaN              None   
55                  -2.448000            -0.0625              None   
56                        NaN                NaN              None   
59                  -0.617500                NaN              None   
6                   -0.484167             0.4675              None   
61                        NaN                NaN              None   
8                   -0.520000             0.1450              None   
9                   -0.780833            -0.1425              None   

            bmi_rate_AB  bmi_rate_BC  bmi_rate_AC  fat_rate_AB  fat_rate_BC  \
subject_id                                                                    
1             -0.590000      -0.1225    -0.382222    -0.820000       -0.850   
10            -0.266667      -0.0875    -0.221875    -0.350000  

In [139]:
none_columns = X.columns[(X.applymap(lambda x: x is None)).any()]
print(none_columns.tolist())

['weight_kg_rate_AC']


In [148]:
# X = X.drop(columns=['weight_kg_rate_AC'])

# 제외할 컬럼 목록
cols_to_remove = [
    'weeks_AB',
    'weight_kg_rate_AB', 'weight_kg_rate_BC', 'weight_kg_rate_AC',
    'bmi_rate_AB', 'bmi_rate_BC', 'bmi_rate_AC',
    'fat_rate_AB', 'fat_rate_BC', 'fat_rate_AC',
    'pure_fat_rate_AB', 'pure_fat_rate_BC', 'pure_fat_rate_AC'
]

# 제외한 y 데이터
y_cleaned = y.drop(columns=cols_to_remove)

In [141]:
X

weight_kg_rate_AB  weight_kg_rate_BC  bmi_rate_AB  bmi_rate_BC  \
subject_id                                                                   
1                   -1.644000            -0.3425    -0.590000      -0.1225   
10                  -0.788333            -0.2550    -0.266667      -0.0875   
11                  -0.750000            -0.1825    -0.231667      -0.0550   
12                  -1.222000             0.2425    -0.450000       0.0900   
13                  -0.926667            -0.0050    -0.325000      -0.0025   
14                  -0.895833             0.1675    -0.260833       0.0500   
18                        NaN                NaN          NaN          NaN   
19                  -0.357500             0.0900    -0.148333       0.0375   
2                   -1.804000            -0.2375    -0.670000      -0.0875   
20                  -0.706667            -0.2400    -0.245833      -0.0850   
21                  -0.982000             0.0600    -0.390000       0.0225   
22                  -1.942000             1.0150    -0.560000       0.2925   
23                  -0.976667            -0.0325    -0.308333      -0.0100   
24                  -0.792500            -0.1575    -0.279167      -0.0550   
25                  -1.576000            -0.1325    -0.512000      -0.0425   
26                  -2.172000            -0.3250    -0.610000      -0.0925   
27                  -1.414000            -0.1375    -0.492000      -0.0475   
28                  -1.358000            -0.5325    -0.508000      -0.2000   
29                  -1.844000            -0.5300    -0.534000      -0.1525   
3                   -2.320000             0.5325    -0.730000       0.1700   
30                  -1.942000             0.1000    -0.558000       0.0275   
31                  -0.411667            -0.2200    -0.152500      -0.0825   
33                  -0.520000            -0.1150    -0.160833      -0.0350   
34                  -0.725833            -0.2000    -0.286667      -0.0800   
35                  -0.560833             0.0875    -0.195000       0.0300   
36                  -0.764167            -0.2950    -0.264167      -0.1025   
37                  -1.556000            -0.1000    -0.550000      -0.0350   
38                  -2.068000            -0.3800    -0.718000      -0.1325   
39                  -0.340000             0.1000    -0.125000       0.0375   
4                   -1.004000             0.0275    -0.424000       0.0125   
40                  -1.440000            -0.3075    -0.578000      -0.1225   
41                  -0.750000             0.0825    -0.284167       0.0325   
42                  -2.538000            -0.0450    -0.780000      -0.0125   
43                  -2.312000            -0.0400    -0.694000      -0.0125   
45                  -1.452000            -0.0875    -0.532000      -0.0325   
46                  -1.884000            -0.5675    -0.626000      -0.1875   
47                  -1.436000            -0.3650    -0.496000      -0.1250   
48                  -0.757500             0.1075    -0.230833       0.0325   
49                  -0.806667            -0.2400    -0.257500      -0.0750   
5                   -2.460000            -0.0875    -0.768000      -0.0275   
50                  -0.365833             0.0150    -0.120000       0.0050   
51                  -0.681667            -0.1600    -0.220000      -0.0525   
52                  -1.988000             0.4525    -0.622000       0.1425   
53                  -0.773333                NaN    -0.228333          NaN   
55                  -2.448000            -0.0625    -0.900000      -0.0225   
56                        NaN                NaN          NaN          NaN   
59                  -0.617500                NaN    -0.217500          NaN   
6                   -0.484167             0.4675    -0.193333       0.1875   
61                        NaN                NaN          NaN          NaN   
8                   -0.520000     

In [165]:
y_cleaned.shape

(53, 12)

In [164]:
y_cleaned

weight_kg_C_A  weight_kg_C_B  weight_kg_B_A  bmi_C_A  bmi_C_B  \
subject_id                                                                  
1                   -9.59          -1.37          -8.22    -3.44    -0.49   
2                   -9.97          -0.95          -9.02    -3.70    -0.35   
3                   -9.47           2.13         -11.60    -2.97     0.68   
4                   -4.91           0.11          -5.02    -2.07     0.05   
5                  -12.65          -0.35         -12.30    -3.95    -0.11   
6                   -3.94           1.87          -5.81    -1.57     0.75   
8                   -5.66           0.58          -6.24    -2.08     0.21   
9                   -9.94          -0.57          -9.37    -3.46    -0.20   
10                 -10.48          -1.02          -9.46    -3.55    -0.35   
11                  -9.73          -0.73          -9.00    -3.00    -0.22   
12                  -5.14           0.97          -6.11    -1.89     0.36   
13                 -11.14          -0.02         -11.12    -3.91    -0.01   
14                 -10.08           0.67         -10.75    -2.93     0.20   
15                    NaN          -0.19            NaN      NaN    -0.07   
18                 -10.42            NaN            NaN    -3.15      NaN   
19                  -3.93           0.36          -4.29    -1.63     0.15   
20                  -9.44          -0.96          -8.48    -3.29    -0.34   
21                  -4.67           0.24          -4.91    -1.86     0.09   
22                  -5.65           4.06          -9.71    -1.63     1.17   
23                 -11.85          -0.13         -11.72    -3.74    -0.04   
24                 -10.14          -0.63          -9.51    -3.57    -0.22   
25                  -8.41          -0.53          -7.88    -2.73    -0.17   
26                 -12.16          -1.30         -10.86    -3.42    -0.37   
27                  -7.62          -0.55          -7.07    -2.65    -0.19   
28                  -8.92          -2.13          -6.79    -3.34    -0.80   
29                 -11.34          -2.12          -9.22    -3.28    -0.61   
30                  -9.31           0.40          -9.71    -2.68     0.11   
31                  -5.82          -0.88          -4.94    -2.16    -0.33   
33                  -6.70          -0.46          -6.24    -2.07    -0.14   
34                  -9.51          -0.80          -8.71    -3.76    -0.32   
35                  -6.38           0.35          -6.73    -2.22     0.12   
36                 -10.35          -1.18          -9.17    -3.58    -0.41   
37                  -8.18          -0.40          -7.78    -2.89    -0.14   
38                 -11.86          -1.52         -10.34    -4.12    -0.53   
39                  -3.68           0.40          -4.08    -1.35     0.15   
40                  -8.43          -1.23          -7.20    -3.38    -0.49   
41                  -8.67           0.33          -9.00    -3.28     0.13   
42                 -12.87          -0.18         -12.69    -3.95    -0.05   
43                 -11.72          -0.16         -11.56    -3.52    -0.05   
45                  -7.61          -0.35          -7.26    -2.79    -0.13   
46                 -11.69          -2.27          -9.42    -3.88    -0.75   
47                  -8.64          -1.46          -7.18    -2.98    -0.50   
48                  -8.66           0.43          -9.09    -2.64     0.13   
49                 -10.64          -0.96          -9.68    -3.39    -0.30   
50                  -4.33           0.06          -4.39    -1.42     0.02   
51                  -8.82          -0.64          -8.18    -2.85    -0.21   
52                  -8.13           1.81          -9.94    -2.54     0.57   
53                    NaN            NaN          -9.28      NaN      NaN   
54                    NaN          -0.41            NaN      NaN    -0.11   
55                 -12.49          -0.25         -12.24    -4.59    -0.09   
56      

In [135]:
y

weight_kg_C_A  weight_kg_C_B  weight_kg_B_A  bmi_C_A  bmi_C_B  \
subject_id                                                                  
1                   -9.59          -1.37          -8.22    -3.44    -0.49   
2                   -9.97          -0.95          -9.02    -3.70    -0.35   
3                   -9.47           2.13         -11.60    -2.97     0.68   
4                   -4.91           0.11          -5.02    -2.07     0.05   
5                  -12.65          -0.35         -12.30    -3.95    -0.11   
6                   -3.94           1.87          -5.81    -1.57     0.75   
8                   -5.66           0.58          -6.24    -2.08     0.21   
9                   -9.94          -0.57          -9.37    -3.46    -0.20   
10                 -10.48          -1.02          -9.46    -3.55    -0.35   
11                  -9.73          -0.73          -9.00    -3.00    -0.22   
12                  -5.14           0.97          -6.11    -1.89     0.36   
13                 -11.14          -0.02         -11.12    -3.91    -0.01   
14                 -10.08           0.67         -10.75    -2.93     0.20   
15                    NaN          -0.19            NaN      NaN    -0.07   
18                 -10.42            NaN            NaN    -3.15      NaN   
19                  -3.93           0.36          -4.29    -1.63     0.15   
20                  -9.44          -0.96          -8.48    -3.29    -0.34   
21                  -4.67           0.24          -4.91    -1.86     0.09   
22                  -5.65           4.06          -9.71    -1.63     1.17   
23                 -11.85          -0.13         -11.72    -3.74    -0.04   
24                 -10.14          -0.63          -9.51    -3.57    -0.22   
25                  -8.41          -0.53          -7.88    -2.73    -0.17   
26                 -12.16          -1.30         -10.86    -3.42    -0.37   
27                  -7.62          -0.55          -7.07    -2.65    -0.19   
28                  -8.92          -2.13          -6.79    -3.34    -0.80   
29                 -11.34          -2.12          -9.22    -3.28    -0.61   
30                  -9.31           0.40          -9.71    -2.68     0.11   
31                  -5.82          -0.88          -4.94    -2.16    -0.33   
33                  -6.70          -0.46          -6.24    -2.07    -0.14   
34                  -9.51          -0.80          -8.71    -3.76    -0.32   
35                  -6.38           0.35          -6.73    -2.22     0.12   
36                 -10.35          -1.18          -9.17    -3.58    -0.41   
37                  -8.18          -0.40          -7.78    -2.89    -0.14   
38                 -11.86          -1.52         -10.34    -4.12    -0.53   
39                  -3.68           0.40          -4.08    -1.35     0.15   
40                  -8.43          -1.23          -7.20    -3.38    -0.49   
41                  -8.67           0.33          -9.00    -3.28     0.13   
42                 -12.87          -0.18         -12.69    -3.95    -0.05   
43                 -11.72          -0.16         -11.56    -3.52    -0.05   
45                  -7.61          -0.35          -7.26    -2.79    -0.13   
46                 -11.69          -2.27          -9.42    -3.88    -0.75   
47                  -8.64          -1.46          -7.18    -2.98    -0.50   
48                  -8.66           0.43          -9.09    -2.64     0.13   
49                 -10.64          -0.96          -9.68    -3.39    -0.30   
50                  -4.33           0.06          -4.39    -1.42     0.02   
51                  -8.82          -0.64          -8.18    -2.85    -0.21   
52                  -8.13           1.81          -9.94    -2.54     0.57   
53                    NaN            NaN          -9.28      NaN      NaN   
54                    NaN          -0.41            NaN      NaN    -0.11   
55                 -12.49          -0.25         -12.24    -4.59    -0.09   
56      

In [166]:
# 공통 인덱스 (subject_id) 추출
common_index = X.index.intersection(y_cleaned.index)

# 공통 인덱스 기준으로 X, y 정렬 및 필터링
X_common = X.loc[common_index]
y_cleaned_common = y_cleaned.loc[common_index]

In [167]:
print(X_common.shape)
print(y_cleaned_common.shape)
print(X_common.index.equals(y_cleaned_common.index))  # True여야 안전

(51, 883)
(51, 12)
True


In [145]:
X_common.columns[:30]

Index(['weight_kg_rate_AB', 'weight_kg_rate_BC', 'bmi_rate_AB', 'bmi_rate_BC',
       'bmi_rate_AC', 'fat_rate_AB', 'fat_rate_BC', 'fat_rate_AC',
       'pure_fat_rate_AB', 'pure_fat_rate_BC', 'pure_fat_rate_AC', 'treatment',
       'sex', 'age_yrs', 'height_cm', 'base_weight_kg', 'base_bmi_kg_m2',
       'base_body_fat_pct', 'base_pure_fat_kg', 'exp_C_A_7897955',
       'exp_C_A_7898858', 'exp_C_A_7899870', 'exp_C_A_7906475',
       'exp_C_A_7916282', 'exp_C_A_7919314', 'exp_C_A_7921821',
       'exp_C_A_7923762', 'exp_C_A_7927926', 'exp_C_A_7930714',
       'exp_C_A_7931951'],
      dtype='object')

In [174]:
# treatment: 'very-low-calorie diet' → True, 나머지 → False
X_common['is_vlcd'] = X_common['treatment'] == 'very-low-calorie diet'

# sex: 'F' → True (여성), 나머지 → False
X_common['is_female'] = X_common['sex'] == 'F'

# 기존 열 제거 (원한다면)
X_common = X_common.drop(columns=['treatment', 'sex'])

In [168]:
y_cleaned_common.columns

Index(['weight_kg_C_A', 'weight_kg_C_B', 'weight_kg_B_A', 'bmi_C_A', 'bmi_C_B',
       'bmi_B_A', 'fat_C_A', 'fat_C_B', 'fat_B_A', 'pure_fat_C_A',
       'pure_fat_C_B', 'pure_fat_B_A'],
      dtype='object')

# **데이터 증강 + PCA + 정규화 + Optuna 튜닝 + CatBoostRegressor 학습**

In [144]:
!pip install catboost

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 99.2/99.2 MB 8.8 MB/s eta 0:00:00


In [169]:
!pip install optuna

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 395.9/395.9 kB 8.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 242.7/242.7 kB 12.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.5/78.5 kB 4.6 MB/s eta 0:00:00


In [170]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from catboost import CatBoostRegressor, Pool
from sklearn.model_selection import train_test_split
import optuna
from sklearn.cluster import KMeans

In [175]:
from sklearn.impute import SimpleImputer

imputer = SimpleImputer(strategy='mean')  # 또는 'median'

# X와 y 각각 결측치 평균 대체
X_imputed = imputer.fit_transform(X_common)
y_imputed = imputer.fit_transform(y_cleaned_common)

# 다시 DataFrame으로 변환 (컬럼 이름 유지)
X_imputed = pd.DataFrame(X_imputed, columns=X_common.columns, index=X_common.index)
y_imputed = pd.DataFrame(y_imputed, columns=y_cleaned_common.columns, index=y_cleaned_common.index)

In [176]:
def augment_data(X, y, n_noise_samples=50, n_kmeans_clusters=10, kmeans_std=0.01, noise_std=0.01):
    # 노이즈 증강
    X_noise = X.sample(n=n_noise_samples, replace=True, random_state=0)
    y_noise = y.loc[X_noise.index]

    noise = np.random.normal(0, noise_std, size=X_noise.shape)
    X_noise_augmented = pd.DataFrame(X_noise.to_numpy() + noise, columns=X.columns)
    y_noise_augmented = y_noise.reset_index(drop=True)

    # KMeans 중심 기반 증강
    from sklearn.cluster import KMeans
    kmeans = KMeans(n_clusters=n_kmeans_clusters, random_state=0, n_init=10)
    kmeans.fit(X_imputed)
    centers = kmeans.cluster_centers_
    noise_center = np.random.normal(0, kmeans_std, size=centers.shape)
    X_kmeans_augmented = pd.DataFrame(centers + noise_center, columns=X.columns)

    # y의 평균값으로 대응 (단순하게 처리)
    y_kmeans_augmented = pd.DataFrame([y.mean()] * len(X_kmeans_augmented), columns=y.columns)

    # 원본 + 증강 합치기
    X_aug = pd.concat([X, X_noise_augmented, X_kmeans_augmented], ignore_index=True)
    y_aug = pd.concat([y.reset_index(drop=True), y_noise_augmented, y_kmeans_augmented], ignore_index=True)

    return X_aug, y_aug

In [177]:
# 증강
X_aug, y_aug = augment_data(X_imputed, y_imputed, n_noise_samples=100, n_kmeans_clusters=10, noise_std=0.01)

# 정규화 및 차원 축소
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X_aug)

pca = PCA(n_components=50)
X_pca = pca.fit_transform(X_scaled)

In [180]:
X_scaled.shape

(161, 883)

In [178]:
X_common.shape

(51, 883)

In [182]:
X_pca.shape

(161, 50)

X_pca에는 원래 입력 데이터 X_commondm feature특징이 각각 얼마나 보존되어 있느냐 확인할 필요 있음.

In [183]:
from sklearn.multioutput import MultiOutputRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from catboost import CatBoostRegressor
import optuna

def objective(trial):
    # CatBoost 개별 모델 파라미터
    params = {
        'depth': trial.suggest_int('depth', 4, 10),
        'learning_rate': trial.suggest_float('learning_rate', 0.01, 0.3),
        'l2_leaf_reg': trial.suggest_float('l2_leaf_reg', 1, 10),
        'iterations': 500,
        'early_stopping_rounds': 20,
        'verbose': 0
    }

    # 데이터 분할
    X_train, X_valid, y_train, y_valid = train_test_split(X_pca, y_aug, test_size=0.2, random_state=42)

    # CatBoost 모델 래핑
    base_model = CatBoostRegressor(**params)
    model = MultiOutputRegressor(base_model)

    # 학습
    model.fit(X_train, y_train)

    # 예측 및 평가 (평균 R² 점수 사용)
    y_pred = model.predict(X_valid)
    score = r2_score(y_valid, y_pred, multioutput='uniform_average')

    return score

In [184]:
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=50)
print("Best trial:", study.best_trial.params)

[I 2025-06-16 21:25:19,757] A new study created in memory with name: no-name-88a1eee4-bbb2-4fdb-a0ae-b1a1009de20a
[I 2025-06-16 21:26:23,454] Trial 0 finished with value: 0.9552555188810213 and parameters: {'depth': 4, 'learning_rate': 0.1579596948853682, 'l2_leaf_reg': 1.4040949822929867}. Best is trial 0 with value: 0.9552555188810213.
[I 2025-06-16 21:27:16,167] Trial 1 finished with value: 0.9375090543595639 and parameters: {'depth': 4, 'learning_rate': 0.017303012294279982, 'l2_leaf_reg': 2.921288368911944}. Best is trial 0 with value: 0.9552555188810213.
[I 2025-06-16 21:28:41,824] Trial 2 finished with value: 0.9643353559313881 and parameters: {'depth': 5, 'learning_rate': 0.1751118741715008, 'l2_leaf_reg': 4.149017344648899}. Best is trial 2 with value: 0.9643353559313881.
[I 2025-06-16 21:32:30,934] Trial 3 finished with value: 0.9499146817998655 and parameters: {'depth': 7, 'learning_rate': 0.24174781648475469, 'l2_leaf_reg': 7.074017733861644}. Best is trial 2 with value: 0.

Best trial: {'depth': 4, 'learning_rate': 0.0419998814673207, 'l2_leaf_reg': 1.4499985542901122}


In [186]:
from catboost import CatBoostRegressor, Pool, cv, CatBoostError
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

# 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(X_pca, y_aug, test_size=0.2, random_state=42)

# 모델 파라미터
best_params = study.best_params
best_params.update({
    'iterations': 500,
    'early_stopping_rounds': 20,
    'verbose': 100,  # 100 iteration마다 로그 출력
    'eval_metric': 'R2'
})

In [187]:
# 개별 타겟별 학습 및 시각화
models = []
evals = []

for i, col in enumerate(y_train.columns):
    print(f"\n[Target {i+1}] {col}")

    model = CatBoostRegressor(**best_params)
    train_pool = Pool(X_train, y_train[col])
    valid_pool = Pool(X_test, y_test[col])

    # 학습 로그 저장
    eval_result = {}
    model.fit(train_pool, eval_set=valid_pool, use_best_model=True, verbose=100, plot=False)

    # 학습 로그 가져오기
    eval_result = model.get_evals_result()

    # 저장
    models.append(model)
    evals.append(eval_result)


[Target 1] weight_kg_C_A
0:	learn: 0.0345636	test: -0.0164280	best: -0.0164280 (0)	total: 12.4ms	remaining: 6.19s
100:	learn: 0.9320663	test: 0.8844844	best: 0.8844844 (100)	total: 1.69s	remaining: 6.68s
200:	learn: 0.9875239	test: 0.9569009	best: 0.9569009 (200)	total: 2.86s	remaining: 4.25s
300:	learn: 0.9976008	test: 0.9745443	best: 0.9745443 (300)	total: 4.11s	remaining: 2.72s
400:	learn: 0.9993533	test: 0.9790287	best: 0.9790287 (400)	total: 5.9s	remaining: 1.46s
499:	learn: 0.9997739	test: 0.9802833	best: 0.9802849 (498)	total: 8.5s	remaining: 0us

bestTest = 0.980284913
bestIteration = 498

Shrink model to first 499 iterations.

[Target 2] weight_kg_C_B
0:	learn: 0.0284236	test: -0.0275381	best: -0.0275381 (0)	total: 6.82ms	remaining: 3.4s
100:	learn: 0.9639017	test: 0.8860149	best: 0.8860149 (100)	total: 1.35s	remaining: 5.33s
200:	learn: 0.9980905	test: 0.9604720	best: 0.9604720 (200)	total: 2.88s	remaining: 4.29s
300:	learn: 0.9996055	test: 0.9659862	best: 0.9659862 (300)	to

In [188]:
from sklearn.metrics import r2_score

test_r2_scores = []

for i, col in enumerate(y_test.columns):
    y_pred = models[i].predict(X_test)
    r2 = r2_score(y_test[col], y_pred)
    print(f"[Target {i+1}] {col} - Test R²: {r2:.4f}")
    test_r2_scores.append(r2)

[Target 1] weight_kg_C_A - Test R²: 0.9803
[Target 2] weight_kg_C_B - Test R²: 0.9668
[Target 3] weight_kg_B_A - Test R²: 0.9800
[Target 4] bmi_C_A - Test R²: 0.9706
[Target 5] bmi_C_B - Test R²: 0.9496
[Target 6] bmi_B_A - Test R²: 0.9859
[Target 7] fat_C_A - Test R²: 0.9672
[Target 8] fat_C_B - Test R²: 0.9811
[Target 9] fat_B_A - Test R²: 0.9540
[Target 10] pure_fat_C_A - Test R²: 0.9667
[Target 11] pure_fat_C_B - Test R²: 0.9764
[Target 12] pure_fat_B_A - Test R²: 0.9650


In [190]:
# 결과 출력
print("Best R2:", study.best_value)
print("Best Params:", study.best_params)

Best R2: 0.9703236903218481
Best Params: {'depth': 4, 'learning_rate': 0.0419998814673207, 'l2_leaf_reg': 1.4499985542901122}


In [191]:
import pandas as pd

# PCA 결과의 차원 수에 맞춰 컬럼 이름 생성
pca_cols = [f'PC{i+1}' for i in range(X_test.shape[1])]
X_test_df = pd.DataFrame(X_test, columns=pca_cols)

# y_pred를 DataFrame으로 생성 (이미 위에서 함)
y_pred_df = pd.DataFrame(
    {col: model.predict(X_test) for col, model in zip(y_test.columns, models)},
    index=y_test.index
)
y_pred_df.columns = [f"{col}_pred" for col in y_test.columns]

# 실제값과 예측값 합치기
comparison_df = pd.concat([X_test_df.reset_index(drop=True),
                           y_test.reset_index(drop=True),
                           y_pred_df.reset_index(drop=True)], axis=1)

# 결과 보기
comparison_df.head()

PC1       PC2        PC3        PC4       PC5        PC6        PC7  \
0  -3.599441 -1.120713  -3.472462  -4.114827 -3.973381   1.344415  -1.618885   
1  -9.973699  7.258163  -5.175429   5.337060  3.171625  -4.559797   3.850432   
2   0.488131 -6.610284   4.630564   2.660737 -1.332226   6.725531   1.073349   
3  21.199219  4.327849  20.307006   2.946912 -9.316966   1.721544 -16.250519   
4  12.780689  7.536796   7.979020 -11.065576  2.522889 -11.906498  -2.238432   

        PC8       PC9      PC10  ...  weight_kg_B_A_pred  bmi_C_A_pred  \
0 -1.598001  3.768998  1.688139  ...           -7.162468     -2.717334   
1 -2.272159  6.162198 -5.778123  ...           -7.281193     -3.008531   
2 -0.026577 -8.065782 -6.368232  ...           -4.619274     -1.441004   
3  3.710147  1.797117  3.265820  ...           -6.087200     -1.988581   
4 -5.645789  7.199335 -8.048962  ...          -12.121348     -3.695888   

   bmi_C_B_pred  bmi_B_A_pred  fat_C_A_pred  fat_C_B_pred  fat_B_A_pred  \
0     -0.110848     -2.616304     -5.185523     -1.660911     -3.397904   
1     -0.497500     -2.490870     -5.176722     -1.147878     -3.840856   
2      0.015607     -1.508138     -1.859231      1.283977     -3.203995   
3      0.323630     -2.309774     -2.706562      1.088915     -3.669675   
4     -0.080569     -3.768540     -8.687589     -0.381251     -8.769691   

   pure_fat_C_A_pred  pure_fat_C_B_pred  pure_fat_B_A_pred  
0          -7.734613          -1.529094          -6.045392  
1          -7.588260          -1.471918          -6.002203  
2          -3.841742           1.337524          -5.255635  
3          -4.950970           1.228633          -6.035281  
4         -10.739087          -0.353081         -10.708030  

[5 rows x 74 columns]

In [193]:
import joblib

# 모델 저장
joblib.dump(model, "/content/drive/MyDrive/BIO_MEDICAL_PROJECT/obesity_gene_analysis/data/my_model.pkl")

['/content/drive/MyDrive/BIO_MEDICAL_PROJECT/obesity_gene_analysis/data/my_model.pkl']

In [195]:
import numpy as np

np.save("/content/drive/MyDrive/BIO_MEDICAL_PROJECT/obesity_gene_analysis/data/X_pca.npy", X_pca)
np.save("/content/drive/MyDrive/BIO_MEDICAL_PROJECT/obesity_gene_analysis/data/y_aug.npy", y_aug)